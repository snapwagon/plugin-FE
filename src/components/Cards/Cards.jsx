import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import toMarkdown from 'to-markdown';
import marked from 'marked';

import { analytics } from '../../utils/utils';

import Image from '../../components/Image/Image';
import Content from '../../components/Content/Content';
import Section from '../../components/Content/Section';
import Button from '../../components/Button/Button';

class Cards extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isFinePrintVisible: false
    };

    this.handleShowFinePrint = this.handleShowFinePrint.bind(this);
    this.handleInterest = this.handleInterest.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleKeyPress);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyPress);
  }

  handleKeyPress(e) {
    // escape
    if (e.keyCode === 27) {
      this.props.handleClose(e);
    // enter
    } else if (e.keyCode === 13) {
      this.handleInterest(e);
    }
  }

  handleInterest(e) {
    analytics.track('Product Viewed', {
      offerId: this.props.offer.id,
      clientId: this.props.clientId
    });
    return this.props.handleContinue(this.props.offer.id);
  }

  handleShowFinePrint(e) {
    this.setState({
      isFinePrintVisible: true
    });
  }

  render() {
    const savingsAmt = this.props.offer.value - this.props.offer.discounted_value;
    const detailLine = `Discount: ${this.props.offer.discount_percentage}% Value: $${this.props.offer.value}`;

    const markup = this.props.offer.desc;
    const sanitize = (htmlString) => {
      return { __html: marked(toMarkdown(htmlString), {sanitize: true})};
    };

    return (
      <div
        className={cx(
          'snapW-Cards',
          `snapW-Cards--${this.props.orientation}`,
          this.props.classNames
        )}
      >
        <Section type="Header">
          <Image src={this.props.offer.image_url} alt={this.props.offer.title} />
        </Section>
        <Section type="Body">
          <Content title={this.props.offer.title} subtitle={`ONLY $${this.props.offer.discounted_value}`} tagline={detailLine} />
          {this.props.offer.desc && (
           !this.state.isFinePrintVisible ?
             (<button
               className={'snapW-SubSection--Link--Button'}
               onClick={this.handleShowFinePrint}
             >
               Fine print...
             </button>) :
             (
               <div
                 dangerouslySetInnerHTML={sanitize(markup)}
                 className={`snapW-SubSection snapW-SubSection--${this.state.isFinePrintVisible}`}
               />
             )
           )
         }
          <Button
            onClick={this.handleInterest}
            size="small"
            text={`Save $${savingsAmt}!`}
          />
        </Section>
      </div>
    );
  }
}

const {
  arrayOf,
  node,
  bool,
  string,
  oneOf,
  oneOfType,
  objectOf,
  func
} = PropTypes;

Cards.propTypes = {
  classNames: oneOfType([
    string,
    objectOf(bool)
  ]),
  orientation: oneOf([
    'column',
    'row'
  ]),
  handleClose: func,
  handleContinue: func,
  offer: objectOf(node)
};

Cards.defaultProps = {
  handleClose() {},
  handleContinue() {},
  classNames: undefined,
  orientation: 'row',
  offer: {},
};

export default Cards;
