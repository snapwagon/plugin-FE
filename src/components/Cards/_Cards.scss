.coup-Cards {
  display: flex;
  flex-wrap: wrap;
  background-color: $white;

  &--column {
    flex-direction: column;
    justify-content: flex-start;
  }

  &--row {
    justify-content: space-around;
  }
}
