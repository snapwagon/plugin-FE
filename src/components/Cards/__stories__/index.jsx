/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import times from 'lodash/times';

import bottleImage from 'images/jump.jpg';
import lifestyleImage from 'images/jump.jpg';

import Cards from '../Cards';
import Image from '../../Image/Image';
import Content from '../../Content/Content';
import Section from '../../Content/Section';
import Button from '../../Button/Button';

storiesOf('Cards', module).addDecorator((story) => (
  <div style={{
    maxWidth: '300px',
    margin: '0 auto'
  }}>
    {story()}
  </div>
)).addWithInfo('rendered as a column of <Cards />', () => {
  return (
    <Cards orientation="column">
      <Section type="Header">
        <div style={{
          margin: '0 auto'
        }}>
          <svg width="80" height="150" viewBox="0 0 33 47" xmlns="http://www.w3.org/2000/svg">
            <path d="M24.153.976a28.237 28.237 0 0 0-3.535-.754l-.467 7.842h-7.76L11.925.222C10.72.402 9.542.654 8.392.98 7.91 5.999 9.522 18.407 0 24.336v16.389s5.781 6.2 16.274 6.2c10.49 0 16.27-6.598 16.27-6.598V24.336c-9.524-5.929-7.912-18.34-8.39-23.36" fill="#0F346C" fillRule="evenodd"/>
          </svg>
        </div>
      </Section>
      <Section type="Body">
        <Content title="Blue Apron" subtitle="Building a Better Food System"/>
      </Section>
    </Cards>
  );
});

storiesOf('Cards', module).addDecorator((story) => (
  <div style={{
    maxWidth: '800px',
    margin: '0 auto'
  }}>
    {story()}
  </div>
)).addWithInfo('rendered as a row of <Card />', () => (
  <Cards>
    <Section type="Header">
      <Image src={lifestyleImage} alt="Jump into Spring!"/>
    </Section>
    <Section type="Body">
      <Content title="ONLY $17.00" subtitle="Jump into June!" tagline="DISCOUNT: 34%"/>
      <Button
        onClick={action('onClick')}
        type="medium"
        customStyle={{ color:"white", background: "#155885" }}
      />
    </Section>

  </Cards>
));

// .add('multiple cards in edit mode', () => {
//   const stubItems = [{
//     title: 'Pinot Noir',
//     subTitle: 'Medel',
//     imageUrl: bottleImage,
//     subText: 'Australia 2014',
//     id: 3
//   },
//   {
//     title: 'Pinot Noir',
//     subTitle: 'Medel',
//     imageUrl: bottleImage,
//     subText: 'Australia 2014',
//     id: 2
//   },
//   {
//     title: 'Pinot Noir',
//     subTitle: 'Medel',
//     imageUrl: bottleImage,
//     subText: 'Australia 2014',
//     id: 8
//   }];

//   const children = stubItems.map((cardData, index) => {
//     return (
//       <Card
//         applyMinimumHeight={false}
//         key={cardData.id}
//         orientation="landscape"
//       >
//         <Section type="Header">
//           <Image
//             alt="Le P'tit Paysan"
//             src={cardData.imageUrl}
//           />
//         </Section>
//         <Section type="Body">
//           <Content
//             title={cardData.title}
//             subtitle={cardData.subTitle}
//             tagline={cardData.subText}
//           />
//           <Button
//             classNames="coup-Content--action"
//             color="blue"
//             text="Select"
//             type="small"
//           />
//         </Section>
//       </Card>
//     );
//   });

//   return (
//     <Cards orientation="column">
//       {children}
//     </Cards>
//   );
// })
// .add('is Selected State', () => {
//   const stubItems = [{
//     title: 'Pinot Noir',
//     subTitle: 'Medel',
//     imageUrl: bottleImage,
//     subText: 'Australia 2014',
//     id: 3
//   },
//   {
//     title: 'Pinot Noir',
//     subTitle: 'Medel',
//     imageUrl: bottleImage,
//     subText: 'Australia 2014',
//     id: 2
//   },
//   {
//     title: 'Pinot Noir',
//     subTitle: 'Medel',
//     imageUrl: bottleImage,
//     subText: 'Australia 2014',
//     id: 8
//   }];

//   const children = stubItems.map((cardData, index) => {
//     return (
//       <Card orientation="landscape" applyMinimumHeight={false} isSelected={true}>
//         <Section type="Header">
//           <Image
//             alt="Le P'tit Paysan"
//             src={cardData.imageUrl}
//           />
//         </Section>
//         <Section type="Body">
//           <Content
//             title={cardData.title}
//             subtitle={cardData.subTitle}
//             tagline={cardData.subText}
//           />
//           <Button classNames="coup-Content--action" color="blue" text="Select" type="small" />
//         </Section>
//       </Card>
//     );
//   });

//   return (
//     <div style={{ width: '300px' }}>
//       <Cards classNames="coup-Cards--column">
//         {children}
//       </Cards>
//     </div>
//   );
// });
