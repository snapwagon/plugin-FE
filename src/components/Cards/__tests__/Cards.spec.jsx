/* eslint-disable no-undef */

import React from 'react';
import renderer from 'react-test-renderer';

import Cards from '../Cards';
import Header from '../../Header/Header.jsx';

describe('<Cards /> snapshots', () => {
  test('rendered with required props', () => {
    const tree = renderer.create(
      <Cards>
        <Header classNames="coup-Cards__header" text="Also Available:" />
        <div>Aye</div>
      </Cards>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('rendered with props', () => {
    const tree = renderer.create(
      <Cards oreientation="column">
        <div>Hello?</div>
      </Cards>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
