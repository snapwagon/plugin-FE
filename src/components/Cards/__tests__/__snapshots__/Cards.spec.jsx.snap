// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cards /> snapshots rendered with props 1`] = `
<div
  className="snapW-Cards snapW-Cards--row"
>
  <div>
    Hello?
  </div>
</div>
`;

exports[`<Cards /> snapshots rendered with required props 1`] = `
<div
  className="snapW-Cards snapW-Cards--row"
>
  <header
    className="snapW-Header snapW-Cards__header"
  >
    Also Available:
  </header>
  <div>
    Aye
  </div>
</div>
`;
